util/wait/wait.go

```diff
@@ -0,0 +1,52 @@
+package wait
+
+import (
+	"errors"
+	"time"
+)
+
+// Backoff contains the configuration for exponential backoff
+type Backoff struct {
+	// InitialDelayInterval represents the initial amount of time of sleep
+	InitialDelayInterval time.Duration
+	// MaxDelayInterval represents the maximal amount of time of sleep between
+	// retries
+	MaxDelayInterval time.Duration
+	// MaxRetryAttempts is the maximal number of retries before exiting with
+	// an error. A value of zero signifies unlimited retry attemps
+	MaxRetryAttempts int
+	// Multiplier is used to increment the current interval by multiplying it with
+	// this multiplier
+	Multiplier float64
+}
+
+// BackoffFunc represents a function to retry with exponential backoff, which
+// returns true if the attempt is successful, or an error if the retry should be
+// aborted
+type BackoffFunc func() (done bool, err error)
+
+// ErrMaxRetryAttempts is returned when the number of maximal retry attempts is
+// reached
+var ErrMaxRetryAttempts = errors.New("maximal number of retry attempts reached")
+
+// ExponentialBackoff retries the provided func with exponential backoff, until
+// the maximal number of retries is reached
+func (b *Backoff) ExponentialBackoff(fn BackoffFunc) error {
+	sleep := b.InitialDelayInterval
+
+	for i := 0; i < b.MaxRetryAttempts || b.MaxRetryAttempts == 0; i++ {
+		if i != 0 {
+			// Sleep for the determined duration
+			time.Sleep(sleep)
+
+			// Exponentially increase that sleep duration
+			sleep = time.Duration(float64(sleep) * b.Multiplier)
+		}
+
+		if ok, err := fn(); err != nil || ok {
```

Comment:
Here I would suggest
```
done, err := fn(i)
if done || err == nil {
    return err
}
if i == MaxRetryAttempts {
    return fmt.Errorf("giving up after %d retries: %s", i, err)
}
```

Then, backoff funcs can return `false, err` when they encounter some error that they want to retry on.

Reply:
I'm not sure to understand what difference it would make. The goal here was to let `BackoffFunc` return the following combinaison:

* `false, nil` - The attempt failed and we should retry, and the error has been optionally logged (e.g. https://github.com/sensu/sensu-go/pull/1057/files#diff-6ea70e0a4de93ebcb83000a028edc163R179)
* `true, nil` - The attempt succeeded, we can immediately exit (e.g. https://github.com/sensu/sensu-go/pull/1057/files#diff-6ea70e0a4de93ebcb83000a028edc163R185)
* `false, err` - A fatal error was encountered during this attempt, we should immediately exit and return the error so the caller can handle the failure